source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,351 @@
+      registerKeyValue(Keys.HEALTH, SpongeHealthData.this::health);
+  }
+
+That's it. The ``DataManipulator`` should be done now.
+
+2. Implement the ImmutableDataManipulator
+=========================================
+
+Implementing the ``ImmutableDataManipulator`` is similar to implementing the mutable one.
+
+The only differences are:
+
+* The class name is formed by prefixing the mutable ``DataManipulator``\ s name with ``Immutable``
+* Inherit from ``ImmutableAbstractData`` instead
+* Instead of ``registerGettersAndSetters()``, the method is called ``registerGetters()``
+
+When creating ``ImmutableDataHolder``\ s or ``ImmutableValue``\ s, check if it makes sense to use the
+``ImmutableDataCachingUtil``. For example if you have ``WetData`` which contains nothing more than a boolean, it
+is more feasible to retain only two cached instances of ``ImmutableWetData`` - one for each possible value. For
+manipulators and values with many possible values (like ``SignData``) however, caching may prove too expensive.
+
+.. tip::
+
+    It may be beneficial to declare the fields of an ``ImmutableDataManipulator`` as ``final`` in order to
+    prevent accidental changes.
+
+3. Register the Key in the KeyRegistry
+======================================
+
+The next step is to register your ``Key``\ s to the ``KeyRegistry``. To do so, locate the
+``org.spongepowered.common.data.key.KeyRegistry`` class and find the static ``registerKeys()`` function.
+There add a line to register (and create) your used keys.
+
+.. code-block:: java
+
+    public static void registerKeys() {
+        keyMap.put("max_health", makeSingleKey(Double.TYPE, MutableBoundedValue.class, of("MaxHealth")));
+    }
+
+The ``keyMap`` maps strings to ``Key``\ s. The string used should be the corresponding constant name from
+the ``Keys`` utility class in lowercase. The ``Key`` itself is created by one of the static methods
+provided by ``KeyFactory``, in most cases ``makeSingleKey``. ``makeSingleKey`` requires first a class reference
+for the underlying data, which in our case is a "Double", then a class reference for the ``Value`` type used.
+The third argument is the ``DataQuery`` used for serialization. It is created from the statically imported
+``DataQuery.of()`` method accepting a string. This string should also be the constant name, stripped of
+underscores and capitalization changed to upper camel case.
+
+.. tip::
+    For primitive types (like double, int, boolean), use the constant ``TYPE`` provided in its wrapper class, not the class reference.
+
+4. Implement the DataProcessor
+==============================
+
+Next up is the ``DataProcessor``. A ``DataProcessor`` serves as a bridge between our ``DataManipulator`` and
+Minecraft's objects. Whenever any data is requested from or offered to ``DataHolders`` that exist in Vanilla
+Minecraft, those calls end up being handled by a ``DataProcessor`` or a ``ValueProcessor``.
```

Comment:
"those calls end up being delegated to a ``DataProcessor`` or a ``ValueProcessor``.